src/main/java/htsjdk/samtools/util/GZIIndex.java

```diff
@@ -0,0 +1,365 @@
+                    throw getCorruptedIndexException(indexPath,
+                            String.format("index entries in misplaced order - %s vs %s",
+                                    entries.get(i - 1), entry));
+                }
+
+                entries.add(entry);
+            }
+
+            return new GZIIndex(entries);
+        }
+    }
+
+    private static final IOException getCorruptedIndexException(final Path indexPath, final String msg) {
+        return new IOException(String.format("Corrupted index file: %s (%s)",
+                msg,
+                indexPath == null ? "unknown" : indexPath.toUri()));
+    }
+
+    /**
+     * Builds a {@link GZIIndex} on the fly from a BGZIP file.
+     *
+     * <p>Note that this method does not write the index on disk. Use {@link #writeIndex(Path)} on
+     * the returned object to safe the index.
+     *
+     * @param bgzipFile the bgzip file.
+     *
+     * @return in-memory .gzi index.
+     *
+     * @throws IOException if an I/O error occurs.
+     */
+    public static final GZIIndex buildIndex(final Path bgzipFile) throws IOException {
+        if (bgzipFile == null) {
+            throw new IllegalArgumentException("null input path");
+        }
+        // open the file for reading as a block-compressed file
+        try (final BlockCompressedInputStream bgzipStream = new BlockCompressedInputStream(Files.newInputStream(bgzipFile))) {
+
+            // store the entries as a list
+            final List<IndexEntry> entries = new ArrayList<>(2);
+            entries.add(FIRST_MAPPING);
+
+            // accumulator for number of bytes read to use in the offset for the index-entry
+            int currentOffset = 0;
+            // until the end of the stream
+            while (bgzipStream.read() != -1) {
+                currentOffset++;
+                // if we are at the end of the block
+                if (bgzipStream.endOfBlock()) {
+                    // gets the block address (compressed offset) - requires to parse with the file pointer utils
+                    final long compressed = BlockCompressedFilePointerUtil.getBlockAddress(bgzipStream.getFilePointer());
+                    // add a new IndexEntry
+                    entries.add(new IndexEntry(compressed, currentOffset));
+                }
+            }
+            // construct by converting into an array
+            return new GZIIndex(entries);
+        }
+    }
+
+    /**
+     * Creates a {@link GZIIndex} from a BGZIP file and store it in memory and disk.
+     *
+     * @param bgzipFile the bgzip file.
+     * @param overwrite if the .fai index already exists override it if {@code true}; otherwise, throws a {@link IOException}.
+     *
+     * @return the in-memory representation for the created index.
+     * @throws IOException  if an IO error occurs.
+     */
+    public static GZIIndex createIndex(final Path bgzipFile, final boolean overwrite) throws IOException {
```

Comment:
I'm not sure this method is necessary, although it could be convenient for someone.  It's not tested right now.

Reply:
I did it in a similar way that the createIndex in the FASTA code; I think that it is nice for creating the default index and do not bother about extensions.

I will add a test afterwards, when I add tests for `java.nio.Path` with Jimfs, because if not it will override the current ones.